elastic/kibana
# [Security Solution][Exceptions][Builder] -Move exception builder entry item exceptions ui over to lists

## Summary

Beginning to move the exceptions UI out of the security solution plugin and into the lists plugin. In order to keep PRs (relatively) small, I plan to move single components at a time. This should also then help more easily pinpoint the source of any issues that come up along the way.

The next couple PRs will focus on the exception builder. This one in particular is focused on moving over the `BuilderEntryItem` which deals with rendering the individual exception item entries. An entry can be of type `match`, `match_any`, `list`, `exists`, or `nested`. The component makes use of the autocomplete fields which use the index patterns to display possible fields and field values. 

One of the decisions made in this PR was to have consumers of the `BuilderEntryItem` pass through the autocomplete service as opposed to the `lists` plugin adding it as a dependency. The reason being that it is likely that plugins using the lists plugin will already be consuming either the data plugin or if alerting takes exceptions in, then they'll be consuming alerting. In an effort to avoid some possible icky circular dependency issues, though it best to make the service passed in, as we had already been doing with the hooks in the `lists` plugin.

Quick Summary:
- `x-pack/plugins/security_solution/public/common/components/exceptions/builder/entry_item.tsx` → ` x-pack/plugins/lists/public/exceptions/components/builder/entry_renderer.tsx`
  - Corresponding unit test file moved as well 
  - Updated security solution exception builder to pull `BuilderEntryItem` from lists plugin
- Copied `x-pack/plugins/security_solution/public/common/components/autocomplete/` to ` x-pack/plugins/lists/public/exceptions/components/autocomplete/`
  - As these components are also used in non exceptions flows in security solution, did not remove there
  - Working on creating an issue to see if these components can be taken up by EUI (reached out to the team and they suggested this as a first step) 
  - Unskips cypress exceptions modal test

##### Entry Item (**_THIS PR DEALS WITH_ 👇🏽**)
<img width="1022" alt="Screen Shot 2021-03-14 at 11 31 04 PM" src="https://user-images.githubusercontent.com/10927944/111112669-c9bbb580-851d-11eb-85e1-5fff829bd4f0.png">

##### Exception Item
<img width="1031" alt="Screen Shot 2021-03-14 at 11 32 15 PM" src="https://user-images.githubusercontent.com/10927944/111112689-d50ee100-851d-11eb-96fc-b32ce51917ea.png">

##### Builder Buttons
<img width="1023" alt="Screen Shot 2021-03-14 at 11 32 39 PM" src="https://user-images.githubusercontent.com/10927944/111112739-e821b100-851d-11eb-8ef5-ac87733d6ef3.png">

##### Exception Item Comments
<img width="1026" alt="Screen Shot 2021-03-14 at 11 33 11 PM" src="https://user-images.githubusercontent.com/10927944/111112766-f7086380-851d-11eb-91f3-d433c0290c6a.png">

##### Builder Modal
<img width="1027" alt="Screen Shot 2021-03-14 at 11 33 30 PM" src="https://user-images.githubusercontent.com/10927944/111112787-02f42580-851e-11eb-95d5-dc968deb0670.png">

### Notes
-  There may still be work to do to get the component working for all use cases, however, for the moment being, concentrating on moving existing functionality over
- The autocomplete components that were moved into the `lists` plugin here I think could live elsewhere. Personally, I think it'd be best if they were components that EUI supported as they are consumed in numerous other places within the security app. To avoid disrupting these other consumers, I did a copy paste of them for now to be used by the `lists` plugin. https://github.com/elastic/kibana/issues/94875

### Testing
Prior to making these changes, added a number of tests to ensure coverage of existing functionality. For manual testing, please checkout PR and play around with the following:
  - Add/edit an exception item
    - if an existing item, make sure that it populates as expected in the modal 
  - Add/edit different entry types like `match`, `match_any`, `nested` etc.
  - Play around with the component in storybook

### Checklist
